ff/src/biginteger/mod.rs

```diff
@@ -421,6 +421,53 @@ impl<const N: usize> BigInteger for BigInt<N> {
         }
     }
 
+    #[inline]
+    fn mul(&self, other: &Self) -> (Self, Self) {
+        if self.is_zero() || other.is_zero() {
+            let zero = Self::zero();
+            return (zero, zero);
+        }
+
+        let mut r = crate::const_helpers::MulBuffer::<N>::zeroed();
+
+        let mut carry = 0;
+
+        for i in 0..N {
+            for j in 0..N {
+                r[i + j] = mac_with_carry!(r[i + j], self.0[i], other.0[j], &mut carry);
+            }
+            r.b1[i] = carry;
+            carry = 0;
+        }
+
+        return (Self(r.b0), Self(r.b1));
+    }
+
+    #[inline]
+    fn mul_low(&mut self, other: &Self) {
+        if self.is_zero() || other.is_zero() {
+            *self = Self::zero();
+            return;
+        }
+
+        let mut res = Self::zero();
+        let mut carry = 0;
+
+        for i in 0..N {
+            for j in 0..(N - i) {
+                res.0[i + j] = mac_with_carry!(res.0[i + j], self.0[i], other.0[j], &mut carry);
+            }
+            carry = 0;
+        }
+
+        *self = res
+    }
```

Comment:
```suggestion
    #[inline]
    fn mul_low(&self, other: &Self) -> Self {
        if self.is_zero() || other.is_zero() {
            *self = Self::zero();
            return;
        }

        let mut res = Self::zero();
        let mut carry = 0;

        for i in 0..N {
            for j in 0..(N - i) {
                res.0[i + j] = mac_with_carry!(res.0[i + j], self.0[i], other.0[j], &mut carry);
            }
            carry = 0;
        }

        res
    }
```

Reply:
(For consistency, the operations should all return values. If we want in place operations, we should mark those explicitly)

Reply:
Sure no problem, I've just followed the convention of other operations here like `mul2` or `muln`.

Just did a follow up PR here: https://github.com/arkworks-rs/algebra/pull/775